evidence_test.go

```diff
@@ -0,0 +1,320 @@
+}
+
+func TestEvidence_FromJSON_invalid_kat(t *testing.T) {
+	tokenBytes, err := os.ReadFile("test/evidence_invalid_kat.json")
+	require.NoError(t, err)
+
+	expectedErr := "validation failed: validation of key attestation token failed: validation of cert & pub info failed: invalid certificate information: no digest information in certify info"
+	e := &Evidence{}
+	err = e.FromJSON(tokenBytes)
+	assert.EqualError(t, err, expectedErr)
+}
+
+func TestEvidence_SetTokens_invalidPat(t *testing.T) {
+	var e Evidence
+	err := e.SetTokens(
+		mustBuildValidKAT(t),
+		buildInValidPAT(t),
+	)
+	expectedErr := "validation failed: validation of platform attestation token failed: missing key identifier"
+	assert.EqualError(t, err, expectedErr)
+
+}
+
+func TestEvidence_SetTokens_invalidKat(t *testing.T) {
+	var e Evidence
+	err := e.SetTokens(
+		buildInValidKAT(t),
+		mustBuildValidPAT(t),
+	)
+	expectedErr := "validation failed: validation of key attestation token failed: missing signature"
+	assert.EqualError(t, err, expectedErr)
+
+}
+func TestEvidence_SetTokens_missingKat(t *testing.T) {
+	var e Evidence
+	err := e.SetTokens(
+		nil,
+		mustBuildValidPAT(t),
+	)
+	expectedErr := "validation failed: nil key attestation token supplied"
+	assert.EqualError(t, err, expectedErr)
+}
+
+func TestEvidence_Verify_missingBoth(t *testing.T) {
+	var e Evidence
+	key := generateTestECDSAKey(t)
+	pk := key.Public().(*ecdsa.PublicKey)
+
+	err := e.Verify(pk)
+	expectedErr := "missing Parsec TPM key attestation token"
+	assert.EqualError(t, err, expectedErr)
+}
+
+func TestEvidence_Verify_missingKat(t *testing.T) {
+	e := &Evidence{Pat: mustBuildValidPAT(t)}
+	key := generateTestECDSAKey(t)
+	pk := key.Public().(*ecdsa.PublicKey)
+
+	err := e.Verify(pk)
+	expectedErr := "missing Parsec TPM key attestation token"
+	assert.EqualError(t, err, expectedErr)
+}
+
+func TestEvidence_Verify_missingPat(t *testing.T) {
+	e := &Evidence{Kat: mustBuildValidKAT(t)}
+	key := generateTestECDSAKey(t)
+	pk := key.Public().(*ecdsa.PublicKey)
+
+	err := e.Verify(pk)
+	expectedErr := "missing Parsec TPM platform attestation token"
+	assert.EqualError(t, err, expectedErr)
+}
+
+func TestEvidence_Verify_ECKey_Failed(t *testing.T) {
+	e := &Evidence{Kat: mustBuildValidKAT(t), Pat: mustBuildValidPAT(t)}
+	key := generateTestECDSAKey(t)
+	pk := key.Public().(*ecdsa.PublicKey)
+
+	err := e.Verify(pk)
+	expectedErr := "failed to verify signature on key attestation token: failed to verify signature: Verification failed"
+	assert.EqualError(t, err, expectedErr)
+
+}
+
+func TestEvidence_Verify_RSAKey_Failed(t *testing.T) {
+	e := &Evidence{Kat: mustBuildValidKAT(t), Pat: mustBuildValidPAT(t)}
+	key := generateTestRSAKey(t)
+	pk := key.Public().(*rsa.PublicKey)
+
+	err := e.Verify(pk)
+	expectedErr := "failed to verify signature on key attestation token: failed to verify signature: invalid public key type: *rsa.PublicKey"
+	assert.EqualError(t, err, expectedErr)
+
+}
```

Comment:
left over to be removed

Reply:
RSA Key is used to generate negative test scenario